File: programming-language/cases/cpp/examples/design-pattern/abstract-factory/abstract-factory.cpp
#include <iostream>

////////////////////////////////////////

class Door {
public:
    virtual void description();
    ~Door() {}
};

class DoorFittingExpert {
public:
    virtual void description();
    ~DoorFittingExpert() {}
};

template<typename Door, typename Expert>
class DoorAbstractFactory {
public:
    Door new_door() { return Door {}; }
    Expert new_expert() { return Expert {}; }
};

////////////////////////////////////////

class WoodenDoor : public Door {
public:
    void description() override { std::cout << "A wooden door!" << '\n'; }
    ~WoodenDoor() {}
};

class IronDoor : public Door {
public:
    void description() override { std::cout << "A iron door!" << '\n'; }
    ~IronDoor() {}
};

////////////////////////////////////////

class Carpenter : public DoorFittingExpert {
public:
    void description() override { std::cout << "I can only fit wooden door!" << '\n'; }
    ~Carpenter() {}
};

class Welder : public DoorFittingExpert {
public:
    void description() override { std::cout << "I can only fit iron door!" << '\n'; }
    ~Welder() {}
};

////////////////////////////////////////

using WoodenDoorFactory = DoorAbstractFactory<WoodenDoor, Carpenter>;
using IronDoorFactory = DoorAbstractFactory<IronDoor, Welder>;


int main() {
    {
        auto factory = WoodenDoorFactory();
        auto door = factory.new_door();
        auto expert = factory.new_expert();
        door.description();
        expert.description();
    }
    {
        auto factory = IronDoorFactory();
        auto door = factory.new_door();
        auto expert = factory.new_expert();
        door.description();
        expert.description();
    }
    {
        auto factory = WoodenDoorFactory();
        auto door   = factory.new_door();
        auto expert = factory.new_expert();
        auto pdoor   = static_cast<Door*>             (&door);
        auto pexpert = static_cast<DoorFittingExpert*>(&expert);
        pdoor->description();
        pexpert->description();
    }
    {
        auto factory = IronDoorFactory();
        auto door   = factory.new_door();
        auto expert = factory.new_expert();
        auto pdoor   = static_cast<Door*>             (&door);
        auto pexpert = static_cast<DoorFittingExpert*>(&expert);
        pdoor->description();
        pexpert->description();
    }
    return 0;
}

File: programming-language/cases/cpp/examples/design-pattern/abstract-factory/abstract-factory.rs
// Rust

////////////////////////////////////////

trait Door {
    fn description(&self);
}

trait DoorFittingExpert {
    fn description(&self);
}

trait DoorAbstractFactory<D: Door + Default, E: DoorFittingExpert + Default> {
    fn new_door(&self) -> D { Default::default() }
    fn new_expert(&self) -> E { Default::default() }
}

////////////////////////////////////////

#[derive(Default)]
struct WoodenDoor;

#[derive(Default)]
struct IronDoor;

impl Door for WoodenDoor {
    fn description(&self) { println!("A wooden door!"); }
}

impl Door for IronDoor {
    fn description(&self) { println!("A iron door!"); }
}

////////////////////////////////////////

#[derive(Default)]
struct Carpenter;

#[derive(Default)]
struct Welder;

impl DoorFittingExpert for Carpenter {
    fn description(&self) { println!("I can only fit wooden door!"); }
}

impl DoorFittingExpert for Welder {
    fn description(&self) { println!("I can only fit iron door!"); }
}

////////////////////////////////////////

struct WoodenDoorFactory;
struct IronDoorFactory;

impl DoorAbstractFactory<WoodenDoor, Carpenter> for WoodenDoorFactory {}
impl DoorAbstractFactory<IronDoor, Welder> for IronDoorFactory {}

fn main() {
    {
        let factory = WoodenDoorFactory{};
        let door = factory.new_door();
        let expert = factory.new_expert();
        door.description();
        expert.description();
    }
    {
        let factory = IronDoorFactory{};
        let door = factory.new_door();
        let expert = factory.new_expert();
        door.description();
        expert.description();
    }
}
